src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,507 @@
+
+void websocket_tracker_connection::on_read(error_code ec, std::size_t /* bytes_read */)
+{
+	COMPLETE_ASYNC("websocket_tracker_connection::on_read");
+	if (ec)
+	{
+		if(ec != websocket::error::closed)
+			fail(operation_t::sock_read, ec);
+		close();
+		return;
+	}
+
+	auto const& buf = m_read_buffer.data();
+
+#ifndef TORRENT_DISABLE_LOGGING
+	std::string str(static_cast<char const*>(buf.data()), buf.size());
+	if (auto cb = requester())
+		cb->debug_log("*** WEBSOCKET_TRACKER_READ [ size: %ld, data: %s ]", long(str.size()), str.c_str());
+#endif
+
+	auto ret = parse_websocket_tracker_response({static_cast<char const*>(buf.data()), long(buf.size())}, ec);
+	if(ec)
+	{
+#ifndef TORRENT_DISABLE_LOGGING
+		if (auto cb = requester())
+		{
+			TORRENT_ASSERT(std::holds_alternative<std::string>(ret));
+			cb->debug_log("*** WEBSOCKET_TRACKER_READ [ ERROR: %s ]", std::get<std::string>(ret).c_str());
+		}
+#endif
+		fail(operation_t::handshake, ec);
+		close();
+		return;
+	}
+
+	TORRENT_ASSERT(std::holds_alternative<websocket_tracker_response>(ret));
+	auto response = std::move(std::get<websocket_tracker_response>(ret));
+
+	std::shared_ptr<request_callback> cb;
+	if (auto cit = m_callbacks.find(response.info_hash); cit != m_callbacks.end())
+		cb = cit->second.lock();
+
+	if (cb)
+	{
+		if (response.offer)
+		{
+			response.offer->answer_callback =
+				[info_hash = response.info_hash
+					, self = shared_from_this()
+					, id = response.offer->id
+					, pid = response.offer->pid]
+				(peer_id const& local_pid
+					, aux::rtc_answer const& answer)
+				{
+					self->queue_answer({std::move(info_hash), std::move(local_pid), std::move(answer)});
+					self->start();
+				};
+
+			cb->on_rtc_offer(*response.offer);
+		}
+
+		if(response.answer)
+		{
+			cb->on_rtc_answer(*response.answer);
+		}
+
+		if(response.resp)
+		{
+			response.resp->interval = std::max(response.resp->interval
+				, seconds32{m_man.settings().get_int(settings_pack::min_websocket_announce_interval)});
+
+			cb->tracker_response(tracker_req(), {}, {}, *response.resp);
+		}
+	}
+	else
+	{
+#ifndef TORRENT_DISABLE_LOGGING
+		if (auto cb_ = requester())
+			cb_->debug_log("*** WEBSOCKET_TRACKER_READ [ warning: no callback for info_hash ]");
+#endif
+		m_callbacks.erase(response.info_hash);
+	}
+
+	do_read();
+}
+
+void websocket_tracker_connection::on_write(error_code const& ec, std::size_t /* bytes_written */)
+{
+	COMPLETE_ASYNC("websocket_tracker_connection::on_write");
+	m_write_data.clear();
+	m_sending = false;
+	if (ec)
+	{
+		fail(operation_t::sock_write, ec);
+		close();
+		return;
+	}
+
+	// Continue sending
+	send_pending();
+}
+
+void websocket_tracker_connection::fail(operation_t op, error_code const& ec)
+{
+	tracker_connection::fail(ec, op, ec.message().c_str(), seconds32{120}, seconds32{120});
+}
+
+TORRENT_EXTRA_EXPORT std::variant<websocket_tracker_response, std::string>
+	parse_websocket_tracker_response(span<char const> message, error_code& ec)
+{
+	try {
```

Comment:
you can actually put `try` on the function body block directly, and the `catch` handlers right after the function body block. That way you save one level on indentation in the function